src/OpenTelemetry.Contrib.Instrumentation.Wcf/TelemetryClientMessageInspector.cs

```diff
@@ -0,0 +1,136 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.ServiceModel;
+using System.ServiceModel.Channels;
+using System.ServiceModel.Dispatcher;
+using OpenTelemetry.Context.Propagation;
+using OpenTelemetry.Contrib.Instrumentation.Wcf.Implementation;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Contrib.Instrumentation.Wcf
+{
+    /// <summary>
+    /// An <see cref="IClientMessageInspector"/> implementation which adds telemetry to outgoing requests.
+    /// </summary>
+    public class TelemetryClientMessageInspector : IClientMessageInspector
+    {
+        /// <inheritdoc/>
+        public object BeforeSendRequest(ref Message request, IClientChannel channel)
+        {
+            try
+            {
+                if (WcfInstrumentationActivitySource.Options == null || WcfInstrumentationActivitySource.Options.OutgoingRequestFilter?.Invoke(request) == false)
+                {
+                    WcfInstrumentationEventSource.Log.RequestIsFilteredOut(request.Headers.Action);
+                    return new State
+                    {
+                        SuppressionScope = this.SuppressDownstreamInstrumentation(),
+                    };
+                }
+            }
+            catch (Exception ex)
+            {
+                WcfInstrumentationEventSource.Log.RequestFilterException(ex);
+                return new State
+                {
+                    SuppressionScope = this.SuppressDownstreamInstrumentation(),
+                };
+            }
+
+            Activity activity = WcfInstrumentationActivitySource.ActivitySource.StartActivity(
+                WcfInstrumentationActivitySource.OutgoingRequestActivityName,
+                ActivityKind.Client);
+
+            IDisposable suppressionScope = this.SuppressDownstreamInstrumentation();
+
+            if (activity != null)
+            {
+                string action = request.Headers.Action;
+                activity.DisplayName = action;
+
+                WcfInstrumentationActivitySource.Options.Propagator.Inject(
+                    new PropagationContext(activity.Context, Baggage.Current),
+                    request,
+                    WcfInstrumentationActivitySource.MessageHeaderValueSetter);
+
+                if (activity.IsAllDataRequested)
+                {
+                    int lastIndex = action.LastIndexOf('/');
+
+                    activity.SetTag(WcfInstrumentationConstants.RpcSystemTag, WcfInstrumentationConstants.WcfSystemValue);
```

Comment:
do you need all these attributes on begin? If some of them can be collected on end - it may be better for user experience (request will start processing faster)

Reply:
I can probably move some or all of them but let me ask, does it really change anything? We block for a few ns/ms before sending the call, or we block for a few ns/ms after the call, either way, our net block time should be about the same from the caller's perspective, no?